### PR TITLE
Add API method for get user by ID of an SSH key

### PR DESCRIPTION
**Pull request by bozaro**
GL_ID environment variable in hooks can contains value like:
- user-:id
- key-:id

Unfortunately, I have not found a way to get information about the user by ID for the SSH key via current API.

### CODE REVIEW

**Comment by bozaro** [comment]
Rebased on master to avoid merge conflicts

**Comment by bozaro** [comment]
https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/1233

**Comment by stanhu** [comment]
@bozaro Thanks, no need to send merge requests to both github.com and gitlab.com. One is sufficient.

@DouweM Is there any reason allowing lookup by key ID doesn't make sense? Someone asked about [this feature here](https://groups.google.com/forum/#!topic/gitlabhq/ht_0G6JcHOc).

**Comment by bozaro** [comment]
> @DouweM Is there any reason allowing lookup by key ID doesn't make sense? Someone asked about this feature here.

In any case, I can get a list of all users consistently exhausting their IDs via current API.

The only new knowledge that I get more - see the number of keys on the user. And to know the identifiers of the keys.

I do not see this possibility of any danger.

**Comment by DouweM** [comment]
- Only admin can get a user's keys via the API currently, so only admin should be able to use this lookup feature.
- I'm wondering if there's a more "Railsy"/"RESTy" way to format the URL. How about `keys/:id/user`?

**Comment by bozaro** [comment]
> Only admin can get a user's keys via the API currently, so only admin should be able to use this lookup feature.

We'll let it go at that

> I'm wondering if there's a more "Railsy"/"RESTy" way to format the URL. How about keys/:id/user?

Pull request updated

**Comment by stanhu** [comment]
Thanks. Can you also add a spec to `lib/api/users_spec.rb`?

**Comment by bozaro** [comment]
> Thanks. Can you also add a spec to lib/api/users_spec.rb?

I'm afraid that I did not have enough skill to do so.

**Comment by DouweM** [comment]
@bozaro I changed my mind a bit :) Ideally, we'd have a `keys/:id` (admin-only) that would return a JSON doc of the key, with the user embedded. That's much better than the weird `keys/:id/user` endpoint. Could you change that, and also move this into a separate `lib/api/keys.rb`? For the spec, you can copy the relevant parts from the existing `users_spec.rb`.

**Comment by bozaro** [comment]
@DouweM 
I will try to make these changes.

I think it will take a couple of days. At the moment, I have no experience of programming in Ruby.
For example, I do not know how to run the spec tests.

**Comment by DouweM** [comment]
@bozaro Cool. You should probably get our [dev kit](https://gitlab.com/gitlab-org/gitlab-development-kit) set up.

**Comment by bozaro** [comment]
Updated. Still without specs.

**Comment by bozaro** [comment]
@DouweM 
Update. Specs added.

Dev kit was extremely useful :+1:

File: lib/api/entities.rb [hunk 1]
```diff
@@ -199,6 +199,11 @@ class SSHKey < Grape::Entity
       expose :id, :title, :key, :created_at
     end
 
+    class SSHKeyWithUser < Grape::Entity
```
Comment (DouweM) [inline lib/api/entities.rb]: If you do `SSHKeyWithUser < SSHKey`, you don't need to repeat the first line of exposing :)

File: lib/api/keys.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    resource :keys do
+      # Get single ssh key by id. Only available to admin users.
+      #
+      # Example Request:
+      #   GET /keys/:id
+      get ":id" do
+        authenticated_as_admin!
+
+        key = Key.find_by(id: params[:id])
```
Comment (DouweM) [inline lib/api/keys.rb]: You can simply do `Key.find(params[:id])`.

File: lib/api/keys.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        not_found!('Key') unless key
```
Comment (DouweM) [inline lib/api/keys.rb]: If you use `find` rather than `find_by`, you don't need this line.

**Comment by bozaro** [comment]
Fix review notes.

File: doc/api/users.md [hunk 4]
```diff
@@ -150,6 +150,51 @@ Parameters:
 }
 ```
 
+### Get SSH key with user by ID of an SSH key
```
Comment (DouweM) [inline doc/api/users.md]: This should probably be moved into a separate api/keys doc since it's no longer about users.

**Comment by DouweM** [comment]
@bozaro Code looks great, thanks!

File: doc/api/users.md (hunk 4 above)
Comment (bozaro) [inline doc/api/users.md]: Moved

**Comment by DouweM** [comment]
@bozaro :+1:

**Comment by bozaro** [comment]
Thanks :)

**Pull request merged**